### PR TITLE
Preflights reporting updates

### PR DESCRIPTION
**Pull request by jgruica**

### CODE REVIEW

File: kotsadm/operator/pkg/client/client.go [hunk 1]
```diff
@@ -331,7 +332,7 @@ func (c *Client) registerHandlers(socketClient *socket.Client) error {
 
 	err = socketClient.On("appInformers", func(h *socket.Channel, args InformRequest) {
 		log.Printf("received an inform event: %#v", args)
-		c.applyAppInformers(args.AppID, args.Informers)
+		c.applyAppInformers(args.AppID, args.Informers, args.Sequence)
```
Comment (salahalsaleh) [inline kotsadm/operator/pkg/client/client.go:335]: would've loved to have the sequence param after the app id param :)
```
c.applyAppInformers(args.AppID, args.Sequence, args.Informers)
```

File: pkg/handlers/deploy.go [hunk 2]
```diff
@@ -85,19 +85,14 @@ func (h *Handler) DeployAppVersion(w http.ResponseWriter, r *http.Request) {
 	// preflights reports
 	if !request.IsAirgap {
 		go func() {
+			isCli := false
```
Comment (salahalsaleh) [inline pkg/handlers/deploy.go:88]: i don't agree with this pattern. this is a handler, and can be called from anywhere. deciding if it's a cli or not should not be made here in the api.

File: pkg/handlers/preflight.go [hunk 3]
```diff
@@ -339,7 +339,8 @@ func (h *Handler) PreflightsReports(w http.ResponseWriter, r *http.Request) {
 	clusterID := downstreams[0].ClusterID
 
 	go func() {
-		if err := reporting.SendPreflightsReportToReplicatedApp(license, foundApp.ID, clusterID, 0, true, ""); err != nil {
+		isCli := false
```
Comment (salahalsaleh) [inline pkg/handlers/preflight.go:342]: same here

File: pkg/handlers/update.go [hunk 4]
```diff
@@ -60,13 +60,15 @@ func (h *Handler) AppUpdateCheck(w http.ResponseWriter, r *http.Request) {
...
-			isUpdate := true
-			err = reporting.SendPreflightInfo(foundApp.ID, int(foundApp.CurrentSequence), skipPreflights, isUpdate)
-			if err != nil {
-				logger.Debugf("failed to update preflights reports: %v", err)
-			}
-		}()
+		if deploy {
+			go func() {
+				isCli := true
+				err = reporting.SendPreflightInfo(foundApp.ID, int(foundApp.CurrentSequence), skipPreflights, isCli)
```
Comment (salahalsaleh) [inline pkg/handlers/update.go:66]: `foundApp.CurrentSequence` is not the latest downloaded sequence here because that happens inside a goroutine in the background and most probably after this executed.

File: pkg/online/online.go [hunk 5]
```diff
@@ -237,12 +237,14 @@ func CreateAppFromOnline(pendingApp *types.PendingApp, upstreamURI string, isAut
...
 
-	go func() {
-		isUpdate := false
-		if err := reporting.SendPreflightInfo(pendingApp.ID, int(newSequence), skipPreflights, isUpdate); err != nil {
-			logger.Debugf("failed to send preflights data to replicated app: %v", err)
-		}
-	}()
+	if isAutomated {
+		go func() {
+			isCli := true
```
Comment (salahalsaleh) [inline pkg/online/online.go:242]: same here

File: pkg/reporting/preflight.go [hunk 6]
```diff
@@ -1,97 +1,138 @@
...
-	kotsv1beta1 "github.com/replicatedhq/kots/kotskinds/apis/kots/v1beta1"
-	downstream "github.com/replicatedhq/kots/pkg/kotsadmdownstream"
-	"github.com/replicatedhq/kots/pkg/logger"
-	"github.com/replicatedhq/kots/pkg/store"
-)
-
-func SendPreflightsReportToReplicatedApp(license *kotsv1beta1.License, appID string, clusterID string, sequence int, skipPreflights bool, installStatus string) error {
-	endpoint := license.Spec.Endpoint
-	if !canReport(endpoint) {
+	import (
```
Comment (salahalsaleh) [inline pkg/reporting/preflight.go:3]: indentation is wrong in this whole file

File: pkg/reporting/preflight.go [hunk 7]
```diff
@@ -1,97 +1,138 @@
...
+					logger.Debugf("failed to get app status: %v", err.Error())
+					return
+				}
+				if s.Sequence == int64(sequence) && s.State == appstatustypes.StateReady {
+					appStatus = s.State
+					break
+				}
+			}
+
 			<-time.After(20 * time.Second)
```
Comment (salahalsaleh) [inline pkg/reporting/preflight.go:86]: why is this needed?

File: pkg/reporting/preflight.go [hunk 8]
```diff
@@ -1,97 +1,138 @@
...
-			currentVersion, err := downstream.GetCurrentVersion(appID, clusterID)
+			currentVersionStatus, err := downstream.GetStatusForVersion(appID, clusterID, int64(sequence))
```
Comment (salahalsaleh) [inline pkg/reporting/preflight.go:87]: would make more sense to have this step after the preflight step

File: pkg/reporting/preflight.go [hunk 9]
```diff
@@ -1,97 +1,138 @@
...
+					if err := json.Unmarshal([]byte(p.Result), &preflightResults); err != nil {
+						logger.Debugf("failed to unmarshal preflight results: %v", err.Error())
+						return
+					}
+					preflightState = getPreflightState(preflightResults)
+					break
+				}
+			}
+
+			if (currentVersionStatus != "" && currentVersionStatus != "deploying") {
```
Comment (salahalsaleh) [inline pkg/reporting/preflight.go:112]: no need for this check

File: pkg/kotsadmdownstream/downstream.go [hunk 10]
```diff
@@ -222,6 +222,26 @@ func GetCurrentVersion(appID string, clusterID string) (*types.DownstreamVersion
 	return v, nil
 }
 
+func GetStatusForVersion(appID string, clusterID string, sequence int64) (string, error) {
+	db := persistence.MustGetPGSession()
+	query := `SELECT
```
Comment (emosbaugh) [inline pkg/kotsadmdownstream/downstream.go:227]: can you fix the indentation in these lines?

File: pkg/reporting/preflight.go [hunk 11]
```diff
@@ -68,30 +69,67 @@ func SendPreflightInfo(appID string, sequence int, isSkipPreflights bool, isUpda
 
 	clusterID := downstreams[0].ClusterID
 
-	if isSkipPreflights || isUpdate {
-		// at this point current version status does not exist so it's
-		// neccessary to create thread to get it after version is deployed
+	go func() {
+		appStatus := appstatustypes.StateMissing
+		for start := time.Now(); time.Since(start) < 20*time.Minute; {
```
Comment (emosbaugh) [inline pkg/reporting/preflight.go:74]: is this now a tight loop? it used to poll every 20 seconds

https://github.com/replicatedhq/kots/pull/1712/files#diff-706f7e57e6f7361d82c9a8f1b0c3ddb08c18e88d3ee384cf20964b0ca86254c2L77

File: pkg/reporting/preflight.go [hunk 12]
```diff
@@ -68,30 +69,67 @@ func SendPreflightInfo(appID string, sequence int, isSkipPreflights bool, isUpda
...
+			}
+		}
 
-		// isUpdate means that it's not on initial install
-		go func() {
-			<-time.After(20 * time.Second)
-			currentVersion, err := downstream.GetCurrentVersion(appID, clusterID)
+		preflightState := ""
+		var preflightResults *troubleshootpreflight.UploadPreflightResults
+		for start := time.Now(); time.Since(start) < 5*time.Minute; {
```
Comment (emosbaugh) [inline pkg/reporting/preflight.go:88]: this is another tight loop

File: pkg/reporting/preflight.go [hunk 13]
```diff
@@ -68,30 +69,67 @@ func SendPreflightInfo(appID string, sequence int, isSkipPreflights bool, isUpda
...
 
-	if isSkipPreflights || isUpdate {
-		// at this point current version status does not exist so it's
-		// neccessary to create thread to get it after version is deployed
+	go func() {
+		appStatus := appstatustypes.StateMissing
+		for start := time.Now(); time.Since(start) < 20*time.Minute; {
+			s, err := store.GetStore().GetAppStatus(appID)
+			if err != nil {
+				logger.Debugf("failed to get app status: %v", err.Error())
```
Comment (emosbaugh) [inline pkg/reporting/preflight.go:77]: I know this isnt new code but shouldnt this be at warn level?

File: pkg/reporting/preflight.go [hunk 14]
```diff
@@ -68,30 +69,67 @@ func SendPreflightInfo(appID string, sequence int, isSkipPreflights bool, isUpda
...
-		go func() {
-			<-time.After(20 * time.Second)
-			currentVersion, err := downstream.GetCurrentVersion(appID, clusterID)
+		preflightState := ""
+		var preflightResults *troubleshootpreflight.UploadPreflightResults
+		for start := time.Now(); time.Since(start) < 5*time.Minute; {
+			p, err := store.GetStore().GetPreflightResults(appID, int64(sequence))
 			if err != nil {
-				logger.Debugf("failed to get current downstream version: %v", err)
+				logger.Debugf("failed to get preflight results: %v", err.Error())
```
Comment (emosbaugh) [inline pkg/reporting/preflight.go:91]: I know this isnt new code but shouldnt this be at warn level?

File: pkg/reporting/preflight.go [hunk 15]
```diff
@@ -68,30 +69,67 @@ func SendPreflightInfo(appID string, sequence int, isSkipPreflights bool, isUpda
...
 				return
 			}
-			if currentVersion.Status != "" && currentVersion.Status != "deploying" {
-				if err := SendPreflightsReportToReplicatedApp(license, appID, clusterID, sequence, isSkipPreflights, currentVersion.Status); err != nil {
-					logger.Debugf("failed to send preflights data to replicated app: %v", err)
+
+			if p.Result != "" {
+				if err := json.Unmarshal([]byte(p.Result), &preflightResults); err != nil {
+					logger.Debugf("failed to unmarshal preflight results: %v", err.Error())
```
Comment (emosbaugh) [inline pkg/reporting/preflight.go:97]: I know this isnt new code but shouldnt this be at warn level?

File: pkg/reporting/preflight.go [hunk 16]
```diff
@@ -68,30 +69,67 @@ func SendPreflightInfo(appID string, sequence int, isSkipPreflights bool, isUpda
...
-		}()
-	} else {
-		if err := SendPreflightsReportToReplicatedApp(license, appID, clusterID, sequence, isSkipPreflights, ""); err != nil {
-			return errors.Wrap(err, "failed to send preflights data to replicated app")
 		}
-	}
 
+		currentVersionStatus, err := downstream.GetStatusForVersion(appID, clusterID, int64(sequence))
+		if err != nil {
+			logger.Debugf("failed to get status for version: %v", err)
```
Comment (emosbaugh) [inline pkg/reporting/preflight.go:107]: I know this isnt new code but shouldnt this be at warn level?

File: pkg/reporting/preflight.go [hunk 17]
```diff
@@ -68,30 +69,67 @@ func SendPreflightInfo(appID string, sequence int, isSkipPreflights bool, isUpda
...
+			return
+		}
+
+		if err := SendPreflightsReportToReplicatedApp(license, appID, clusterID, sequence, isSkipPreflights, currentVersionStatus, isCli, preflightState, string(appStatus)); err != nil {
+			logger.Debugf("failed to send preflights data to replicated app: %v", err)
```
Comment (emosbaugh) [inline pkg/reporting/preflight.go:112]: I know this isnt new code but shouldnt this be at warn level?

File: pkg/store/kotsstore/appstatus_store.go [hunk 18]
```diff
@@ -13,26 +13,29 @@ import (
 
 func (s *KOTSStore) GetAppStatus(appID string) (*appstatustypes.AppStatus, error) {
 	db := persistence.MustGetPGSession()
-	query := `select resource_states, updated_at from app_status where app_id = $1`
+	query := `select resource_states, updated_at, sequence from app_status where app_id = $1`
 	row := db.QueryRow(query, appID)
-
+
```
Comment (emosbaugh) [inline pkg/store/kotsstore/appstatus_store.go:18]: need to run go fmt

**Review by emosbaugh — CHANGES_REQUESTED** [review]

File: pkg/reporting/preflight.go (hunk 11 above)
Comment (salahalsaleh) [reply pkg/reporting/preflight.go:74]: this didn't use to run every 20 seconds. it was just a one time hard 20 seconds wait.

Comment (salahalsaleh) [reply pkg/reporting/preflight.go:74]: and i think tight loops are fine here since the code will have to wait for the app status anyways before reporting, right?

Comment (salahalsaleh) [reply pkg/reporting/preflight.go:74]: and it breaks once the app becomes ready here https://github.com/replicatedhq/kots/pull/1712/files#diff-706f7e57e6f7361d82c9a8f1b0c3ddb08c18e88d3ee384cf20964b0ca86254c2R81

Comment (emosbaugh) [reply pkg/reporting/preflight.go:74]: But wont it starve everything else of cpu? or is there a sleep somewhere im not seeing?

Comment (emosbaugh) [reply pkg/reporting/preflight.go:74]: Also its querying the db in GetAppStatus

Comment (salahalsaleh) [reply pkg/reporting/preflight.go:74]: oh, good call, i didn't notice there's no sleep.

File: pkg/reporting/preflight.go (hunk 13 above)
Comment (jgruica) [reply pkg/reporting/preflight.go:77]: What should be warn level debug?
I'm seeing only Debug, Error, Debugf and Info

File: pkg/reporting/preflight.go (hunk 11 above)
Comment (jgruica) [reply pkg/reporting/preflight.go:74]: resolved, added sleep

File: pkg/reporting/preflight.go (hunk 12 above)
Comment (jgruica) [reply pkg/reporting/preflight.go:89]: added sleep

File: pkg/reporting/preflight.go (hunk 13 above)
Comment (emosbaugh) [reply pkg/reporting/preflight.go:77]: I'll let @salahalsaleh chime in here. I just assume these are real errors and they are being masked cause they are debug level. What is the reason for ignoring them? This is old code so not a blocker for merging.

**Review by emosbaugh — APPROVED** [review]

File: pkg/handlers/deploy.go [hunk 19]
```diff
@@ -43,6 +43,7 @@ type DeployAppVersionRequest struct {
 	IsSkipPreflights             bool `json:"isSkipPreflights"`
 	ContinueWithFailedPreflights bool `json:"continueWithFailedPreflights"`
 	IsAirgap                     bool `json:"isAirgap"`
+	IsCLI                        bool `json:"isCLI"`
```
Comment (emosbaugh) [inline pkg/handlers/deploy.go:46]: technically id make this

```go
IsCLI bool `json:"isCli"`
```

File: pkg/reporting/preflight.go (hunk 13 above)
Comment (salahalsaleh) [reply pkg/reporting/preflight.go:77]: i remember the product decision was that we shouldn't log "reporting/metrics" errors at error level

File: pkg/handlers/deploy.go (hunk 19 above)
Comment (salahalsaleh) [reply pkg/handlers/deploy.go:46]: @jgruica +1 on this ^

File: pkg/reporting/preflight.go [hunk 20]
```diff
@@ -2,57 +2,58 @@ package reporting
...
 	if !canReport(endpoint) {
 		return nil
 	}
 
 	urlValues := url.Values{}
-
 	sequenceToStr := fmt.Sprintf("%d", sequence)
 	skipPreflightsToStr := fmt.Sprintf("%t", skipPreflights)
-
+	isCLIToStr := fmt.Sprintf("%t", isCLI)
```
Comment (salahalsaleh) [inline pkg/reporting/preflight.go:29]: would be nice to just have something like
```
urlValues.Set("isCLI", fmt.Sprintf("%t", isCLI))
```

**Review by salahalsaleh — APPROVED** [review]

**Review by salahalsaleh — APPROVED** [review]
